SPICE-0013: Bytes standard library class

bioball opened the pull request:
This adds a new SPICE for working with binary data.

Reference implementation: https://github.com/apple/pkl/pull/1019

netvl commented on spices/SPICE-0013-bytes-standard-library.adoc:
```diff
@@ -0,0 +1,376 @@
+  /// The value represented by these bytes, in hexadecimal encoding.
+  external hex: String
+
+  /// The [MD5](https://en.wikipedia.org/wiki/MD5) hash of these bytes as hexadecimal string.
+  ///
+  /// MD5 is cryptographically broken and should not be used for secure applications.
+  external md5: String
+
+  /// The [SHA-1](https://en.wikipedia.org/wiki/SHA-1) hash of these bytes as hexadecimal string.
+  ///
+  /// SHA-1 is cryptographically broken and should not be used for secure applications.
+  external sha1: String
```
I think that the very wide consensus in IT sphere currently is that MD5 and SHA-1 hashes should not be used anymore, so adding them to the new API seems a bit fishy. What is the intended value of having them, especially MD5?

Maybe instead of providing hashes directly on the class level (which is, admittedly, convenient but which is also pretty inflexible) it would make sense to provide a carefully designed module specifically for hashing, which would provide hashing functions instead? It feels like it would be more flexible and consistent, especially if that module will grow with new hash algorithms in future.

bioball replied:
> Maybe instead of providing hashes directly on the class level (which is, admittedly, convenient but which is also pretty inflexible) it would make sense to provide a carefully designed module specifically for hashing, which would provide hashing functions instead? It feels like it would be more flexible and consistent, especially if that module will grow with new hash algorithms in future.

We can maybe add a `pkl:crypto` module or something similar in the future, if the need arises. But, these methods are also included because we're migrating the existing `Resource.{sha1,sha256,etc}` methods into `Bytes`.

HT154 commented on spices/SPICE-0013-bytes-standard-library.adoc:
```diff
@@ -0,0 +1,376 @@
+----
+@Since { version = "0.29.0" }
+external class Bytes extends Any {
+  /// The value of these bytes.
+  external value: List<UInt8>
+
```
I see reference to `toList()` below. What's the difference between `toList()` and `value`?

Also, would it make sense to add a `length` property?

bioball replied:
Ah, I went back and forth on these two names, and forgot to update this.

I think `toList()` is probably better here; more consistent with the rest of our API.

bioball replied:
I actually went back on this; I'm keeping this as `value`. This better captures the idea that `Bytes` is just a wrapper, and has an underlying value.

bioball replied:
Update: this got changed back again; `toList()` is back. See https://github.com/apple/pkl-evolution/blob/509c11a8b678f3d6f7bd637cbd4456dba107e5d3/spices/SPICE-0013-bytes-standard-library.adoc#bytes-as-a-wrapper for rationale

HT154 commented on spices/SPICE-0013-bytes-standard-library.adoc:
```diff
@@ -0,0 +1,376 @@
+bytes2 = Bytes("AQIDBA==")
+----
+
+==== Concatenating two `Bytes`
+
+[source,pkl]
+----
+bytes3: Bytes = bytes1 + bytes2
+----
```
What about subscript operations? I'd generally expect `bytes1[0]` to return the first byte in the buffer

bioball replied:
The sentiment of this design is to provide a way to view into a byte array, not really to manipulate them. That's also why this class is called `Bytes`, rather than `ByteList` or something like that.

If you need to manipulate them, you can convert to list, before going back to `Bytes`.

We can explore making `List<UInt8>` more efficient for this.

KushalP replied:
As called out here, this functionality exists and is present in the implementation PR: https://github.com/apple/pkl/pull/1019#discussion_r1991869339

> If you need to manipulate them, you can convert to list, before going back to Bytes.

If the desire is to guide users towards this behaviour, should we disable this functionality in `Bytes` or leave it as it is? I'm less concerned with whether subscript operations should/shouldn't exist and what the supported interface looks like for `Bytes` going forwards.

bioball replied:
Right; I will get rid of that behavior in accordance with this SPICE, as long as this behavior is agreed upon.

bioball replied:
Update here: I've updated this SPICE to make `Bytes` itself a sequences of `UInt8`, and added subscript (see [comment](https://github.com/apple/pkl-evolution/pull/14#issuecomment-2870818716))

The actual changes made can be seen in this commit: https://github.com/apple/pkl-evolution/pull/14/commits/2c712eae4eb8d7220d61179f0a8bf85fce6cdb87

HT154 commented on spices/SPICE-0013-bytes-standard-library.adoc:
```diff
@@ -0,0 +1,376 @@
+bytes3: Bytes = bytes1 + bytes2
+----
+
+==== Modifying bytes
+
+NOTE: Although this is possible, Pkl is not efficient at manipulating byte arrays. If heavy computing is needed here, Pkl is likely not the correct tool for this, and should be implemented in terms of `read()` with a custom module reader.
+
+[source,pkl]
+----
+bytes1: Bytes
+
+bytes2 = Bytes(bytes1.toList().take(4))
+----
```
At least one usage I have in mind relies pretty heavily on mutation. Whether it makes sense to build API for this into `Bytes` really depends on the performance of round-tripping through `List`. If that's near-free I think this is fine as-is, but if there is meaningful overhead it'd be great to have some basic mutation APIs.

If there were APIs built for this, I'd like to see something like these:
```pkl
function set(index: UInt, value: UInt8): Bytes
function setRange(start: UInt, end: UInt, value: UInt8): Bytes
function replaceRange(start: UInt, end: UInt, data: Bytes(length == end - start)|List<UInt8>(length == end - start)): Bytes
function map(transform: (Element) -> Result): Bytes
function mapIndexed(transform: (Int, Element) -> Result): Bytes
function mapRange(start: UInt, end: UInt, transform: (Element) -> Result): Bytes
function mapRangeIndexed(start: UInt, end: UInt, transform: (Int, Element) -> Result): Bytes
```

bioball replied:
This is sentiment is "I want to use `Bytes` as a list", which, I think, let's have that convo in the thread here: https://github.com/apple/pkl-evolution/pull/14#discussion_r1984168905

BTW: the signature of the map methods are off. Should be either:

```pkl
function map(transform: (Element) -> Result): List<Result>
function mapIndexed(transform: (Int, Element) -> Result): List<Result>
function mapRange(start: UInt, end: UInt, transform: (Element) -> Result): List<UInt8|Result>
function mapRangeIndexed(start: UInt, end: UInt, transform: (Int, Element) -> Result): List<UInt8|Result>
```

Or:

```pkl
function map(transform: (UInt8) -> UInt8): Bytes
function mapIndexed(transform: (Int, UInt8) -> UInt8): Bytes
function mapRange(start: UInt, end: UInt, transform: (Element) -> UInt8): Bytes
function mapRangeIndexed(start: UInt, end: UInt, transform: (Int, Element) -> UInt8): Bytes
```

Which one did you mean?

HT154 replied:
Ah, I meant the latter. Got distracted after a copy pasta and forgot to edit those.

bioball replied:
Adding a note here: we haven't included these methods in 0.29,. I'm not opposed to it, but we'd like to see real-world perf issues first.

For now, you'd go t o `List` if transformations are needed.

stackoverflow commented on spices/SPICE-0013-bytes-standard-library.adoc:
```diff
@@ -0,0 +1,420 @@
+
+`Bytes` is a wrapper class around an underlying list of bytes.
+
+The underlying bytes can be manipulated by accessing its `value`.
```
Why have a `value` property instead of a `toList()` function?
I imagine `value` will create a new VmList either way.
It's also an easier mnemonic to remember when you forgot how to convert to list.

bioball replied:
I've now renamed this to `toList()`

stackoverflow commented on spices/SPICE-0013-bytes-standard-library.adoc:
```diff
@@ -0,0 +1,426 @@
+  /// The textual rendered output.
+  text: String = renderer.renderDocument(value)
+
+  /// The underlying byte array of [text], in UTF-8 encoding.
```
```suggestion
  /// Defaults to the underlying byte array of [text], in UTF-8 encoding.
```
You can overwrite it, so the doc is misleading.

bioball commented:
I've updated this document quite a bit.

Initially, I went down the direction of simply making `Bytes` a wrapper class, where access and iteration came from converting it into a list.

However, this was somewhat problematic; rationale detailed in https://github.com/apple/pkl-evolution/pull/14/files#diff-3a286f2e59f0c6b7edd5d791874536f6d1c289dd2f0a9aaf82b9c687db6e83bdR594

I've now updated, so that `Bytes` is itself a sequence of bytes.

stackoverflow commented on spices/SPICE-0013-bytes-standard-library.adoc:
```diff
@@ -0,0 +1,643 @@
+
+=== Concatenation
+
+Two `Bytes` values can be concatenated using `+`, producing a the combined result of both values.
```
```suggestion
Two `Bytes` values can be concatenated using `+`, producing the combined result of both values.
```

stackoverflow commented on spices/SPICE-0013-bytes-standard-library.adoc:
```diff
@@ -0,0 +1,643 @@
+
+=== Generators
+
+`Bytes` can be iterated over by a for-generator, and spread syntax.
```
I wonder how useful is that. One could always call `toList()` before spreading or foreaching. It's definitely faster this way, though.

stackoverflow commented on spices/SPICE-0013-bytes-standard-library.adoc:
```diff
@@ -0,0 +1,643 @@
+bytes1 = Bytes(1, 2, 3, 4)
+
+// From a base64-encoded string
+bytes2 = "AQIDBA==".base64DecodedBytes
```
Where is `base64DecodedBytes` defined? It doesn't exist in the current `String` class and you didn't specify it as a new property.

bioball replied:
Whoops, added: https://github.com/apple/pkl-evolution/blob/509c11a8b678f3d6f7bd637cbd4456dba107e5d3/spices/SPICE-0013-bytes-standard-library.adoc#string-class

stackoverflow reviewed (approved):
Look goods. Left some small comments.
